fix(css): `.css?url` support

sapphi-red opened the pull request:
<!-- Thank you for contributing! -->

### Description
- `.css?url` now returns a URL to transformed CSS
  - before, it was returning a URL to transformed CSS in dev, and to raw CSS after build
- ~~`.css?raw&url` now returns a URL to raw CSS~~
  - ~~note that `.css?raw` returns a JS module (`export default ".foo {}"`) but `.css?raw&url` returns a URL to CSS file (`.foo {}`)~~
- `.module.css?url` now throws an error
  - this was never useful because there's no way to get the class name map (`{ 'foo': 'foo_hash' }`) without injecting the actual CSS

fixes #2522
fixes #7842
fixes #13416

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md), especially the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Update the corresponding documentation if needed.
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

sapphi-red committed:
refactor: extract `createSerialPromiseQueue`

sapphi-red committed:
fix(css): `.css?url`/`.css?raw&url` support

sapphi-red committed:
fix(css): throw error when ?url is used with CSS Modules

bluwy commented:
> * `.css?raw&url` now returns a URL to raw CSS

Is there a reason this is supported? I can't quite think of a case where a user would use this, using plain `.css?raw` and maybe `manualChunks` should be good enough. Was it for compat for `.css?url` in prod before?

sapphi-red commented:
> > * `.css?raw&url` now returns a URL to raw CSS
> 
> Is there a reason this is supported? I can't quite think of a case where a user would use this, using plain `.css?raw` and maybe `manualChunks` should be good enough. Was it for compat for `.css?url` in prod before?

I added it for compat for `.css?url` in prod before. I'm fine removing it.

sapphi-red committed:
refactor: remove `.css?raw&url`

sapphi-red committed:
chore: merge main

sapphi-red commented:
In the last meeting, we discussed that we should add `?raw&url` after we find a use case.

---

I removed the `?raw&url` feature and rebased on main branch.

patak-cat reviewed (dismissed):
Nuxt fail seems unrelated. Awesome work @sapphi-red 💚

patak-cat committed:
chore: merge main

josephmarkus commented:
There appears to be a similar issue with SCSS. Say I have:

```jsx
import selectStyles from "somePackage/Select.scss?inline";

export const Theme = () => <style>{selectStyles}</style>;
```

And that `Select.scss` has

```scss
:root,
:host {
  --select-icon: url("./chevron.svg");
}

.selectStyles {
  background-image: var(--select-icon);
}
```

Vite thinks that `./chevron.svg` is relative to the app and not `node_modules` and so tries to resolve it by looking up `/assets/chevron.svg`

patak-dev commented:
Hey @josephmarkus! Your comment in this merged PR will be quickly lost in the repo activity. Would you create a new issue with a minimal reproduction against the latest Vite so we can properly track the problem you are reporting?

josephmarkus commented:
Was just filling in a bug report and whilst reproducing the error realised that my SCSS file imports another SCSS file, but if I target the end file directly, Vite resolves everything fine, so maybe not an issue 🤷‍♂️
